style: Do not immediately convert absolute specified lengths

tomhoule opened the pull request:
<!-- Please describe your changes on the following line: -->

This PR aims to solve issue #15729. I tried to follow the recommendations there as much as possible.

This is my first attempt at contributing to Servo, so this will probably need a lot of input, although I'm eager to make it as polished as possible.

### Current status

- The base inaccuracy issue seems solved, as can be easily verified with the `console.log` based example in the issue.
- Very basic unit tests were added.

### Input needed

I have doubts mainly about the right way to represent these new enum variants for the various length units:

1. With new enum variants in `NoCalcLength` *and* newtypes (current solution)
2. With a `NoCalcLength::Absolute` variant that contains a new `AbsoluteLength` enum, but without newtypes
3. Same as solution 2 but with newtypes

### Work in progress / unresolved issues

- I mostly cared about unit tests until now but will investigate other types of tests
- Tests to check the clamping
- Write a proper commit message

Thanks for your time and feedback :)

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #15729.

<!-- Either: -->
- [X] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16229)
<!-- Reviewable:end -->

tomhoule committed:
Style: Do not immediately convert absolute specified lengths

The NoCalcLength::Absolute variant has been rewritten to accommodate
other units than Au with the new AbsoluteLength enum. This avoids loss
of precision for some operations.

The conversion from floating point absolute lengths to integer
application unit values adopts the same clamping limits as Gecko.

bholley commented:
Awesome, thank you!

r? @emilio per his offer in #15729

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -222,14 +222,96 @@ impl CharacterWidth {
-    Absolute(Au),  // application units
+    /// An absolute length in application units
+    /// See https://drafts.csswg.org/css-values/#absolute-length
+    Au(Au),  // application units
```
So, I think it's worth to keep the distinction between `NoCalcLength::Absolute`, and perhaps have an `AbsoluteLength` enum that represents it, similar to `FontRelativeLength`, wdyt?

emilio replied:
Alternatively we could also give up and inline `FontRelativeLength` here, but unless we have a good reason to, I'd prefer not.

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -222,14 +222,96 @@ impl CharacterWidth {
+
+/// A wrapper representing a floating-point point (pt) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Pt(pub f32);
```
Let's use `CSSFloat` here. They're aliases now, but worth to be consistent.

Also, probably implementing `ToComputedValue` would be a bit better than `From<Au>`, but not necessarily worth I guess.

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -222,14 +222,96 @@ impl CharacterWidth {
+const ABSOLUTE_LENGTH_MIN: f64 = - (1 << 30) as f64;
+
+macro_rules! impl_from_au {
+    ($ty: ty, $multiplier: expr) => {
```
perhaps `$au_per_unit` is a slightly better variable name?

emilio reviewed (commented):
Quite nice! This would also fix a bunch of serialization problems with lengths we currently have, so it's definitely a win, thanks for working on this!

I think this can land with the following nits addressed and the new `AbsoluteLength` enum.

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 272ea6616b05797ede040df88e5cb61abde94a2b with merge aa08ae5ec4dbd20e4663ee3e0dedaadfff86acf3...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/6482)

jdm commented:
From `./mach build-geckolib`:
```
error: `#[derive]` for custom traits is not stable enough for use. It is deprecated and will be removed in v1.15 (see issue #29644)
   --> /home/servo/buildbot/slave/linux-dev/build/components/style/values/specified/length.rs:244:41
    |
244 | #[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
    |                                         ^^^^^^^^^^

error: `#[derive]` for custom traits is not stable enough for use. It is deprecated and will be removed in v1.15 (see issue #29644)
   --> /home/servo/buildbot/slave/linux-dev/build/components/style/values/specified/length.rs:250:41
    |
250 | #[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
    |                                         ^^^^^^^^^^

error: `#[derive]` for custom traits is not stable enough for use. It is deprecated and will be removed in v1.15 (see issue #29644)
   --> /home/servo/buildbot/slave/linux-dev/build/components/style/values/specified/length.rs:256:41
    |
256 | #[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
    |                                         ^^^^^^^^^^

error: `#[derive]` for custom traits is not stable enough for use. It is deprecated and will be removed in v1.15 (see issue #29644)
   --> /home/servo/buildbot/slave/linux-dev/build/components/style/values/specified/length.rs:262:41
    |
262 | #[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
    |                                         ^^^^^^^^^^

error: `#[derive]` for custom traits is not stable enough for use. It is deprecated and will be removed in v1.15 (see issue #29644)
   --> /home/servo/buildbot/slave/linux-dev/build/components/style/values/specified/length.rs:268:41
    |
268 | #[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
    |                                         ^^^^^^^^^^

error: `#[derive]` for custom traits is not stable enough for use. It is deprecated and will be removed in v1.15 (see issue #29644)
   --> /home/servo/buildbot/slave/linux-dev/build/components/style/values/specified/length.rs:274:41
    |
274 | #[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
    |                                         ^^^^^^^^^^

error: `#[derive]` for custom traits is not stable enough for use. It is deprecated and will be removed in v1.15 (see issue #29644)
   --> /home/servo/buildbot/slave/linux-dev/build/components/style/values/specified/length.rs:280:41
    |
280 | #[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
    |                                         ^^^^^^^^^^

error: aborting due to 7 previous errors
```

emilio commented:
Yup, you need to `#[cfg_attr(feature = "servo", derive(HeapSizeOf)]` instead of just deriving it.

tomhoule commented:
Thanks for the feedback!

I made the following changes to address the various issues:

- Changed the various absolute units from newtypes to variants in an `AbsoluteLength` enum, which makes it much more consistent with the rest of the variants in `NoCalcLength`. I kept The `Au` variant as an `Au` value rather than an `i32` because of the convenience methods it has, like `.to_f32_px()`. It's not clear to me if it's worth losing consistency for.
- Replaced `f32` in definitions by `CSSFloat`
- Moved the conversion to application units to a helper function and added a `ToComputedValue` impl to `AbsoluteLength`.
- Moved the `derive(HeapSizeOf)` as advised

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -222,14 +222,128 @@ impl CharacterWidth {
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum AbsoluteLength {
+    /// An absolute length in application units
+    Au(Au),
```
I don't think it makes a lot of sense to keep `AbsoluteLength::Au` around at all, does it?

tomhoule replied:
Indeed, it avoided conversions but it's redundant.

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -311,25 +425,28 @@ impl NoCalcLength {
     /// Returns a `zero` length.
     pub fn zero() -> NoCalcLength {
-        NoCalcLength::Absolute(Au(0))
+        NoCalcLength::Absolute(AbsoluteLength::Au(Au(0)))
```
This should become `AbsoluteLength::Px(0.)`.

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -758,8 +875,8 @@ impl CalcLengthOrPercentage {
-                SimplifiedValueNode::Length(NoCalcLength::Absolute(Au(au))) =>
-                    absolute = Some(absolute.unwrap_or(0) + au),
+                SimplifiedValueNode::Length(NoCalcLength::Absolute(length)) =>
+                    absolute = Some(absolute.unwrap_or(0) + Au::from(length).0),
```
Let's just convert everything to the same unit in `calc()`, (pixels seems adequate), and making `absolute` get a floating point value. You can then do:

```rust
absolute = Some(absoulte.unwrap_or(0.) + Au::from(length).to_f32_px())
```

tomhoule replied:
Done for the scope of the function, did you mean convert it also in the definition for `CalcLengthOrPercentage`?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16283) made this pull request unmergeable. Please resolve the merge conflicts.

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -222,14 +222,122 @@ impl CharacterWidth {
 
+/// Same as Gecko
+const ABSOLUTE_LENGTH_MAX: f64 = (1 << 30) as f64;
+const ABSOLUTE_LENGTH_MIN: f64 = - (1 << 30) as f64;
```
These should be integers, and we should clamp the `Au` value once rounded.

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -222,14 +222,122 @@ impl CharacterWidth {
+/// Helper to convert a floating point length to application units
+fn to_au_round(length: CSSFloat, au_per_unit: CSSFloat) -> Au {
+    Au(
+        ((length as f64) * (au_per_unit as f64))
```
Is there a particular reason you're converting to `f64` here? Shouldn't be needed.

tomhoule replied:
Purely because it was in the sample in the issue, with the bounds we have there seems to be no need indeed.

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -222,14 +222,122 @@ impl CharacterWidth {
+impl AbsoluteLength {
+    fn is_zero(&self) -> bool {
+        match *self {
+             AbsoluteLength::Px(0.)
```
nit: This line should be one space to the left.

emilio reviewed (approved):
Awesome! r=me with these last three bits.

tomhoule commented:
I modified the function that does the clamping to first round and convert to i32, then clamp.

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -220,14 +221,119 @@ impl CharacterWidth {
+/// Helper to convert a floating point length to application units
+fn to_au_round(length: CSSFloat, au_per_unit: CSSFloat) -> Au {
+    let au = (length * au_per_unit).round() as i32;
+    let clamped_au = min(max(au, ABSOLUTE_LENGTH_MIN), ABSOLUTE_LENGTH_MAX);
```
Actually, I was wrong on this (sorry!). This needs to be clamped _before_ rounding and converting to integer, to avoid overflowing on the line before.

tomhoule replied:
No problem, I had doubts and should have voiced them :) Just pushed again. Here is the way I implemented it (without casting to f64 like before):

```rust
fn to_au_round(length: CSSFloat, au_per_unit: CSSFloat) -> Au {
    Au(
        (length * au_per_unit)
        .min(ABSOLUTE_LENGTH_MAX as f32)
        .max(ABSOLUTE_LENGTH_MIN as f32)
        .round() as i32
    )
}
```

emilio commented:
Looks great, thanks!

@bors-servo r+

bors-servo commented:
:clipboard: Looks like this PR is still in progress, ignoring approval

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit d63c35a has been approved by `emilio`

<!-- @bors-servo r=emilio d63c35a92a35e9b07862d4acce62c8ba3ee8e826 -->

bors-servo commented:
:hourglass: Testing commit d63c35a92a35e9b07862d4acce62c8ba3ee8e826 with merge 5ec8c35dd31ed6e684e523c9018d4d7f8f32b444...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3338)

jdm commented:
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/calc.html:
  │ FAIL [expected PASS] calc(1px + 1pt + 1pc + 1in + 1cm + 1mm)
  │   → assert_equals: expected "calc(155.88333333333333px)" but got "calc(155.91666666666666px)"
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/calc.html:61:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ @http://web-platform.test:8000/_mozilla/mozilla/calc.html:59:5
  └ @http://web-platform.test:8000/_mozilla/mozilla/calc.html:58:1
```

emilio commented:
Let's update that test to use the new value, since it relied on rounding differences.

Manishearth commented:
(Stylo trypush @ https://treeherder.mozilla.org/#/jobs?repo=try&revision=ce40864646209bef14531e30df84135715e3b133 , bunch of nice test passes!)

tomhoule commented:
Done :) Sorry for the delay.

Manishearth commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit e4234d0 has been approved by `emilio`

<!-- @bors-servo r=emilio e4234d0a27cdc97554834c915de133415e3e8b23 -->

bors-servo commented:
:hourglass: Testing commit e4234d0a27cdc97554834c915de133415e3e8b23 with merge 0d8ff3ef60579754283d7b306ffde976cec4e14b...

Manishearth commented:
Relevant test expectation updates:

<details>
```patch
diff --git a/layout/base/crashtests/crashtests.list b/layout/base/crashtests/crashtests.list
index 709dcae..692dec1 100644
--- a/layout/base/crashtests/crashtests.list
+++ b/layout/base/crashtests/crashtests.list
@@ -44,7 +44,7 @@ load 265973-1.html
 asserts(6-12) asserts-if(stylo,0) load 265986-1.html # Bug 512405
 load 265999-1.html
 load 266222-1.html
-asserts(1-7) asserts-if(stylo,0) load 266360-1.html # bug 576358
+asserts(1-7) load 266360-1.html # bug 576358
 load 266445-1.html
 asserts(2) asserts-if(stylo,0) load 266445-2.html
 load 268157-1.html
@@ -368,7 +368,7 @@ load 579655.html
 load 580129-1.html
 load 580494-1.html
 load 580834-1.xhtml
-asserts-if(stylo,25) load 589787.html # bug 1324665
+load 589787.html
 load 591075-1.html
 load 591998-1.html
 load 595039-1.html
diff --git a/layout/generic/crashtests/crashtests.list b/layout/generic/crashtests/crashtests.list
index 9283d2f..7ff42bb 100644
--- a/layout/generic/crashtests/crashtests.list
+++ b/layout/generic/crashtests/crashtests.list
@@ -8,7 +8,7 @@ load 255982-2.html
 load 255982-3.html
 load 255982-4.html
 load 264937-1.html
-asserts-if(stylo,4) load 265867-1.html # bug 1324703
+load 265867-1.html
 load 265867-2.html
 load 286491.html
 load 289864-1.html
@@ -297,8 +297,8 @@ load 444861-1.html # bug 1323652
 load 445288.html
 load 448903-1.html
 load 448996-1.html # bug 1323652
-asserts-if(stylo,4) load 451315-1.html # bug 1324672
-asserts-if(stylo,1) load 451317-1.html # bug 1324665
+load 451315-1.html
+load 451317-1.html
 load 451334-1.html
 load 452157-1.html
 load 452157-2.html
@@ -332,7 +332,7 @@ load 468771-1.xhtml
 load 468771-2.xhtml
 load 469859-1.xhtml # bug 1323665
 load 472587-1.xhtml
-asserts-if(stylo,2) load 472617-1.xhtml # bug 1324665
+load 472617-1.xhtml
 load 472774-1.html
 load 472776-1.html
 load 472950-1.html
@@ -350,7 +350,7 @@ load 480345-1.html
 load 481921.html
 load 489462-1.html
 load 489477.html
-asserts-if(stylo,7) load 489480-1.xhtml # bug 1330260
+load 489480-1.xhtml
 load 489647-1.html
 load 493111-1.html
 load 493118-1.html
@@ -363,7 +363,7 @@ load 495875-1.html
 load 495875-2.html
 load 496742.html
 load 499138.html
-asserts-if(stylo,1) load 499857-1.html # bug 1324665
+load 499857-1.html
 load 499862-1.html
 asserts(0-3) load 499885-1.xhtml # Bug 1220265
 load 501535-1.html
@@ -376,7 +376,7 @@ load 508816-1.xul
 load 508908-1.html
 load 509749-1.html
 load 511482.html
-asserts-if(stylo,2) load 512724-1.html # bug 1330259
+load 512724-1.html
 asserts-if(stylo,2) load 512725-1.html # bug 1324640
 load 512749-1.html
 load 513110-1.html
@@ -406,7 +406,7 @@ load 545571-1.html
 load 547338.xul
 load 547843-1.xhtml
 load 551635-1.html
-asserts-if(stylo,4) load 553504-1.xhtml # bug 1324703
+load 553504-1.xhtml
 load 564368-1.xhtml
 load 564968.xhtml
 load 569193-1.html
@@ -558,7 +558,7 @@ load 914891.html
 load 915475.xhtml
 load 927558.html
 load 943509-1.html
-asserts(2-8) asserts-if(stylo,0) load 944909-1.html
+asserts(2-8) load 944909-1.html
 load 946167-1.html
 load 947158.html
 load 949932.html
@@ -571,8 +571,8 @@ load 1001233.html
 load 1001258-1.html
 load 1003441.xul
 pref(layout.css.grid.enabled,true) load 1015562.html
-asserts(1-2) asserts-if(stylo,0) load 1015563-1.html
-asserts(1-2) asserts-if(stylo,0) load 1015563-2.html
+asserts(1-2) load 1015563-1.html
+asserts(1-2) load 1015563-2.html
 asserts(0-300) load 1015844.html # bug 574889
 pref(font.size.inflation.minTwips,200) load 1032450.html
 load 1032613-1.svg
@@ -606,7 +606,7 @@ load 1222783.xhtml
 load 1223568-1.html
 load 1223568-2.html
 load 1224230-1.html
-asserts-if(stylo,4) pref(layout.css.grid.enabled,true) load 1225118.html # bug 1324703
+pref(layout.css.grid.enabled,true) load 1225118.html
 pref(layout.css.grid.enabled,true) load 1225376.html
 pref(layout.css.grid.enabled,true) load 1225592.html
 load 1229437-1.html
@@ -631,7 +631,7 @@ load text-overflow-bug671796.xhtml
 load text-overflow-bug713610.html
 load text-overflow-form-elements.html
 load text-overflow-iframe.html
-asserts-if(Android,2-4) asserts-if(!Android,4) asserts-if(stylo,0) load 1225005.html # bug 682647 and bug 448083
+asserts-if(Android,2-4) asserts-if(!Android,4) load 1225005.html # bug 682647 and bug 448083
 load 1233191.html
 asserts-if(stylo,0-15) load 1271765.html # bug 1324684
 asserts(2) load 1272983-1.html # bug 586628
diff --git a/layout/reftests/backgrounds/reftest-stylo.list b/layout/reftests/backgrounds/reftest-stylo.list
index 2360e34..68a8028 100644
--- a/layout/reftests/backgrounds/reftest-stylo.list
+++ b/layout/reftests/backgrounds/reftest-stylo.list
@@ -112,14 +112,14 @@ pref(layout.css.box-decoration-break.enabled,true) == background-size-cover-clon
 pref(layout.css.box-decoration-break.enabled,true) == background-size-cover-slice.html background-size-cover-slice.html
 
 fails == background-size-monster-ch.html background-size-monster-ch.html
-fails == background-size-monster-cm.html background-size-monster-cm.html
+== background-size-monster-cm.html background-size-monster-cm.html
 fails == background-size-monster-em.html background-size-monster-em.html
 fails == background-size-monster-ex.html background-size-monster-ex.html
-fails == background-size-monster-inches.html background-size-monster-inches.html
-fails == background-size-monster-mm.html background-size-monster-mm.html
-fails == background-size-monster-pc.html background-size-monster-pc.html
-fails == background-size-monster-pt.html background-size-monster-pt.html
-fails == background-size-monster-px.html background-size-monster-px.html
+== background-size-monster-inches.html background-size-monster-inches.html
+== background-size-monster-mm.html background-size-monster-mm.html
+== background-size-monster-pc.html background-size-monster-pc.html
+== background-size-monster-pt.html background-size-monster-pt.html
+== background-size-monster-px.html background-size-monster-px.html
 fails == background-size-monster-rem.html background-size-monster-rem.html
 
 # There seems to be a pixel-snapping problem here, where the repeated background
diff --git a/layout/reftests/border-dotted/reftest-stylo.list b/layout/reftests/border-dotted/reftest-stylo.list
index 4c9e47b..48278ed 100644
--- a/layout/reftests/border-dotted/reftest-stylo.list
+++ b/layout/reftests/border-dotted/reftest-stylo.list
@@ -9,6 +9,6 @@ fuzzy(80,1500) == border-dotted-radius.html border-dotted-radius.html
 fuzzy(80,1500) == border-dashed-no-radius.html border-dashed-no-radius.html
 fuzzy(80,1500) == border-dashed-radius.html border-dashed-radius.html
 fuzzy(80,1500) == border-mixed.html border-mixed.html
-fails == border-dashed-non-integer.html border-dashed-non-integer.html
+== border-dashed-non-integer.html border-dashed-non-integer.html
 fails == border-dashed-radius-zoom.html border-dashed-radius-zoom.html
 fuzzy(80,1500) == border-dotted-radius-zero.html border-dotted-radius-zero.html
diff --git a/layout/reftests/border-radius/reftest-stylo.list b/layout/reftests/border-radius/reftest-stylo.list
index aa3c2cb..bd61578 100644
--- a/layout/reftests/border-radius/reftest-stylo.list
+++ b/layout/reftests/border-radius/reftest-stylo.list
@@ -76,7 +76,7 @@ fails == scrollbar-clamping-1.html scrollbar-clamping-1.html
 fails == scrollbar-clamping-2.html scrollbar-clamping-2.html
 
 # Test for bad corner joins.
-fails == corner-joins-1.xhtml corner-joins-1.xhtml
+== corner-joins-1.xhtml corner-joins-1.xhtml
 == corner-joins-2.xhtml corner-joins-2.xhtml
 
 == scroll-1.html scroll-1.html
diff --git a/layout/reftests/css-gradients/reftest-stylo.list b/layout/reftests/css-gradients/reftest-stylo.list
index 4973cc6..9a8969d 100644
--- a/layout/reftests/css-gradients/reftest-stylo.list
+++ b/layout/reftests/css-gradients/reftest-stylo.list
@@ -68,7 +68,7 @@ fuzzy(3,7860) fuzzy-if(cocoaWidget,5,89041) fuzzy-if(azureSkiaGL,4,90000) == rad
 fails == radial-2e.html radial-2e.html
 fails == radial-2f.html radial-2f.html
 == radial-position-1a.html radial-position-1a.html
-fails == radial-position-1b.html radial-position-1b.html
+== radial-position-1b.html radial-position-1b.html
 fuzzy-if(cocoaWidget,4,22317) fuzzy-if(Android,8,771) == radial-shape-closest-corner-1a.html radial-shape-closest-corner-1a.html
 fuzzy(1,238) fuzzy-if(cocoaWidget,4,22608) fuzzy-if((/^Windows\x20NT\x2010\.0/.test(http.oscpu)||/^Windows\x20NT\x206\./.test(http.oscpu))&&d2d,1,336) fuzzy-if(Android,8,787) fuzzy-if(skiaContent,2,300) == radial-shape-closest-corner-1b.html radial-shape-closest-corner-1b.html
 fails == radial-shape-closest-corner-1c.html radial-shape-closest-corner-1c.html
diff --git a/layout/style/test/stylo-failures.md b/layout/style/test/stylo-failures.md
index 699d32f..47f9a6e 100644
--- a/layout/style/test/stylo-failures.md
+++ b/layout/style/test/stylo-failures.md
@@ -68,7 +68,6 @@ to mochitest command.
   * @keyframes
     * test_keyframes_rules.html [1]
     * test_rules_out_of_sheets.html [1]
-    * test_bug887741_at-rules_in_declaration_lists.html `rules[2]` [1]
   * @support
     * test_supports_rules.html [1]
 * test_box_size_keywords.html: moz-prefixed intrinsic size keyword value [64]
@@ -219,8 +218,6 @@ to mochitest command.
     * test_value_storage.html `on 'stroke` [6]
     * test_compute_data_with_start_struct.html `initial and other values of stroke-dasharray are different` [2]
 * Properties implemented but not in geckolib:
-  * clip property: servo/servo#15729
-    * test_value_storage.html `should be idempotent for 'clip` [4]
   * font-feature-settings property servo/servo#15975
     * test_compute_data_with_start_struct.html `font-feature-settings` [2]
     * test_inherit_computation.html `font-feature-settings` [8]
@@ -244,8 +241,6 @@ to mochitest command.
     * ... `justify-` [6]
     * test_value_storage.html `align-` [9]
     * ... `justify-` [14]
-* @page should ignore properties with viewport units bug 1353191
-  * test_page_parser.html `2v` [8]
 * Stylesheet cloning is somehow busted bug 1348481
   * test_selectors.html `cloned correctly` [157]
   * ... `matched clone` [204]
@@ -306,6 +301,8 @@ to mochitest command.
     * test_logical_properties.html `sideways` [1224]
     * test_value_storage.html `writing-mode` [8]
 * Incorrect parsing
+  * Incorrect bounds
+    * test_bug664955.html `font size is larger than max font size` [2]
   * calc() doesn't support dividing expression servo/servo#15192
     * test_value_storage.html `calc(50px/` [7]
     * ... `calc(2em / ` [9]
@@ -331,8 +328,8 @@ to mochitest command.
 * Incorrect serialization
   * border-radius and -moz-outline-radius shorthand servo/servo#15169
     * test_priority_preservation.html `border-radius` [4]
-    * test_value_storage.html `border-radius:` [92]
-    * ... `-moz-outline-radius:` [45]
+    * test_value_storage.html `border-radius:` [64]
+    * ... `-moz-outline-radius:` [31]
     * test_shorthand_property_getters.html `should condense to shortest possible` [6]
   * transform property servo/servo#15194
     * test_specified_value_serialization.html `bug-721136` [26]
@@ -371,10 +368,6 @@ to mochitest command.
   * test_shorthand_property_getters.html `font shorthand` [1]
   * test_system_font_serialization.html [10]
 * test_value_storage.html `font-size: calc(`: clamp negative value servo/servo#15296 [3]
-* rounding issue
-  * test_value_storage.html `33.5833px` [2]
-  * ... `0.766667px` [2]
-  * ... `105.333px` [2]
 * test_value_storage.html `: var(--a)`: extra whitespace is added for shorthand with variables servo/servo#15295 [*]
 * Negative value should be rejected
   * test_property_syntax_errors.html `transition-duration`: servo/servo#15343 [20]
@@ -393,7 +386,6 @@ to mochitest command.
     * ... `rect(1, ` [2]
 * test_pseudoelement_parsing.html: support parsing some pseudo-classes on some pseudo-elements [5]
 * Unit should be preserved after parsing servo/servo#15346
-  * test_units_length.html [5]
   * test_units_time.html [1]
   * test_page_parser.html `192px` [8]
   * test_bug887741_at-rules_in_declaration_lists.html `188.96666px` [1]
@@ -431,7 +423,6 @@ to mochitest command.
 ## Unknown / Unsure
 
 * test_additional_sheets.html: one sub-test cascade order is wrong [1]
-* test_flexbox_layout.html: resolved width doesn't match expectation [5]
 * test_selectors.html `:nth-child`: &lt;an+b&gt; parsing difference [14]
 * test_selectors_on_anonymous_content.html: xbl and :nth-child [1]
 * test_parse_rule.html `rgb(0, 128, 0)`: color properties not getting computed [6]
diff --git a/layout/svg/crashtests/crashtests.list b/layout/svg/crashtests/crashtests.list
index cc6dfbc..6c22870 100644
--- a/layout/svg/crashtests/crashtests.list
+++ b/layout/svg/crashtests/crashtests.list
@@ -105,7 +105,7 @@ load 587336-1.html
 load 590291-1.svg
 load 601999-1.html
 load 605626-1.svg
-asserts-if(!stylo,2) load 606914.xhtml # bug 606914, bug 1340561 tracks the stylo bit
+asserts(2) load 606914.xhtml # bug 606914, bug 1340561 tracks the stylo bit
 load 610594-1.html
 load 610954-1.html
 load 612662-1.svg
diff --git a/layout/tables/crashtests/crashtests.list b/layout/tables/crashtests/crashtests.list
index 5a478b9..49370d3 100644
--- a/layout/tables/crashtests/crashtests.list
+++ b/layout/tables/crashtests/crashtests.list
@@ -105,7 +105,7 @@ load 416845-1.xhtml
 load 416845-2.xhtml
 load 416845-3.html
 load 420242-1.xhtml
-asserts(8) asserts-if(stylo,4) load 420654-1.xhtml # bug 458238, bug 436123, bug 457397 # bug 1324682
+asserts(8) load 420654-1.xhtml # bug 458238, bug 436123, bug 457397 # bug 1324682
 load 423514-1.xhtml
 load 430374.html
 load 444431-1.html
 ```
</details>

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3335)

emilio commented:
Nice, more passes!

You'll need to mark the following tests as passing (removing the relevant failure lines from the `.ini` file):

```
  ▶ PASS [expected FAIL] /css-values-3_dev/html/min-width-001.htm

  ▶ PASS [expected FAIL] /css-values-3_dev/html/vh_not_refreshing_on_chrome.htm

  ▶ PASS [expected FAIL] /mediaqueries-3_dev/html/min-width-001.htm

  ▶ Unexpected subtest result in /cssom-1_dev/html/cssstyledeclaration-csstext.htm:
  └ PASS [expected FAIL] cssText order

  ▶ Unexpected subtest result in /cssom-1_dev/html/cssstyledeclaration-csstext.htm:
  └ PASS [expected FAIL] another cssText order (non-alphabetical order)

  ▶ Unexpected subtest result in /cssom-1_dev/html/cssstyledeclaration-csstext.htm:
  └ PASS [expected FAIL] whitespaces in value

  ▶ Unexpected subtest result in /cssom-1_dev/html/cssstyledeclaration-csstext.htm:
  └ PASS [expected FAIL] invalid property does not appear
```

jdm commented:
Ignore /css-values-3_dev/html/vh_not_refreshing_on_chrome.htm; that's a known intermittent pass.

Manishearth commented:
@bors-servo r=emilio

(I updated the expectations myself and pushed them, since this is blocking something else and also we need to land this when someone is around to fixup gecko expectations.)

bors-servo commented:
:pushpin: Commit 162fe32 has been approved by `emilio`

<!-- @bors-servo r=emilio 162fe32da5ddbbac49376f045b30c0cb508ae6b3 -->

Manishearth commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 7ecee05 has been approved by `emilio`

<!-- @bors-servo r=emilio 7ecee05e4a098788fcb5f2764914411ed8d0ad54 -->

Manishearth commented:
@bors-servo p=2

needs gecko fixes

bors-servo commented:
:hourglass: Testing commit 7ecee05e4a098788fcb5f2764914411ed8d0ad54 with merge 080d69caa7f0d7dc223c910607fc350fc17ddbfe...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3366)

Manishearth commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit 7ecee05e4a098788fcb5f2764914411ed8d0ad54 with merge c8cd70f333d41864ff7bf7dce84301509c71d1d6...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6258), [arm32](http://build.servo.org/builders/arm32/builds/6476), [arm64](http://build.servo.org/builders/arm64/builds/6510), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6611), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3346), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3369), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6615), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6650), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3301), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3372), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/1787)
Approved by: emilio
Pushing c8cd70f333d41864ff7bf7dce84301509c71d1d6 to master...
